### PR TITLE
[OTX] Add SupCon for semantic segmentation

### PR DESCRIPTION
**Pull request by sungchul1**
# This PR includes 
- Add supcon for semantic segmentation modules, configs
- Enable detcon when training w/ fp16
- Update selfsl & supcon data_pipeline to use `DefaultFormatBundle`

## TODOs

- [x] experiments for reproducibility and improvement
- [x] change target branch from `feature/selfsl/cls` to `feature/otx` after merging https://github.com/openvinotoolkit/training_extensions/pull/1406
- [ ] ~unit test~ 
→ It will be implemented in other PR because this PR is so huge and it's unit test depends on self-sl for semantic segmentation.
- [x] integration test
`tests/integration/cli/segmentation/test_segmentation.py::TestSegmentationCLI::test_otx_train_supcon`

## To be considered in the future
- If with incremental setting, `model.num_classes` is automatically set from `256` to `len(model_classes)`. if not, `model.num_classes` remains `256`. https://github.com/openvinotoolkit/training_extensions/blob/cc643d3dd4ab9f8811d1a8e9a77959b05725b318/otx/mpa/seg/stage.py#L136-L138

### CODE REVIEW

**Comment by sungchul1** [comment]
- [x] `2023-02-01T04:41:22.7597434Z FAILED tests/integration/cli/classification/test_classification.py::TestMultiClassClassificationCLI::test_otx_train_selfsl[Custom_Image_Classification_EfficinetNet-B0]`
- [ ] `2023-02-01T04:41:22.7598667Z FAILED tests/integration/cli/detection/test_instance_segmentation.py::TestInstanceSegmentationCLI::test_nncf_optimize[Custom_Counting_Instance_Segmentation_MaskRCNN_EfficientNetB2B]`
- [x] `2023-02-01T04:41:22.7600116Z FAILED tests/integration/cli/segmentation/test_segmentation.py::TestSegmentationCLI::test_otx_train_selfsl[Custom_Semantic_Segmentation_Lite-HRNet-18-mod2_OCR]`

File: otx/algorithms/classification/tasks/inference.py [hunk 1]
```diff
@@ -328,7 +328,7 @@ def _init_recipe_hparam(self) -> dict:
                 ConfigDict(warmup_iters=warmup_iters) if warmup_iters > 0 else ConfigDict(warmup_iters=0, warmup=None)
             )
 
-        if params.enable_early_stopping:
+        if params.enable_early_stopping and self._recipe_cfg.get("evaluation", None):  # type: ignore
```
Comment (sungmanc) [inline otx/algorithms/classification/tasks/inference.py]: Why use `# type: ignore`? to avoid mypy?, If so, it's not desirable.

If this is for mypy, how about using below?
`if params.enable_early_stopping is not None and self._recipe_cfg.get("evaluation", None) is not None:`

File: otx/algorithms/common/adapters/mmcv/hooks.py [hunk 2]
```diff
@@ -666,3 +666,84 @@ class ForceTrainModeHook(Hook):
     def before_train_epoch(self, runner):
         """Make sure to put a model in a training mode before train epoch."""
         runner.model.train()
+
+
+@HOOKS.register_module()
+class TwoCropTransformHook(Hook):
```
Comment (sungmanc) [inline otx/algorithms/common/adapters/mmcv/hooks.py:672]: Just for curiosity, as far as I remember, the similar function was also implemented for Self-SL classification. Is it wrong? I just want to avoid the duplicate implementation

Comment (sungmanc) [reply otx/algorithms/common/adapters/mmcv/hooks.py:672]: Could you convert the doc-string format as Google's one? We'll use this format as standard doc-string format

Ref.
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: otx/algorithms/common/adapters/mmcv/hooks.py [hunk 3]
```diff
@@ -666,3 +666,84 @@ class ForceTrainModeHook(Hook):
...
+            # for RepeatDataset
+            dataset = runner.data_loader.dataset.dataset
+        else:
+            dataset = runner.data_loader.dataset
+
+        return dataset
+
+    # pylint: disable=inconsistent-return-statements
+    @check_input_parameters_type()
+    def _find_two_crop_transform(self, transforms):
```
Comment (sungmanc) [inline otx/algorithms/common/adapters/mmcv/hooks.py]: Need type hint

**Comment by sungchul1** [comment]
`FAILED tests/integration/cli/detection/test_instance_segmentation.py::TestInstanceSegmentationCLI::test_nncf_optimize[Custom_Counting_Instance_Segmentation_MaskRCNN_EfficientNetB2B]`
Is there any way to solve this fail?

**Comment by sungmanc** [comment]
> `FAILED tests/integration/cli/detection/test_instance_segmentation.py::TestInstanceSegmentationCLI::test_nncf_optimize[Custom_Counting_Instance_Segmentation_MaskRCNN_EfficientNetB2B]` Is there any way to solve this fail?

References:
https://github.com/openvinotoolkit/training_extensions/pull/1574#issuecomment-1413321980

According to the @cih9088 's opinion, This FAIL only occurred in CI. could you check this? @yunchu

**Comment by cih9088** [comment]
> According to the @cih9088 's opinion, This FAIL only occurred in CI. could you check this? @yunchu

Can I access the CI server to debug this or at least container image for CI? ~~It happens in Geti as well and a somewhat related ticket has been filed. (CVS-102888)~~

File: otx/algorithms/common/tasks/training_base.py [hunk 4]
```diff
@@ -57,7 +57,7 @@
 
 logger = get_logger()
 TRAIN_TYPE_DIR_PATH = {
-    TrainType.INCREMENTAL.name: ".",
+    TrainType.INCREMENTAL.name: "",
```
Comment (jaegukhyun) [inline otx/algorithms/common/tasks/training_base.py]: @harimkang Is it okay to change this. AFAIK there is specific reason for this from building side

File: otx/algorithms/segmentation/tasks/inference.py [hunk 5]
```diff
@@ -187,7 +187,12 @@ def _init_recipe(self):
         logger.info(f"train type = {self._train_type}")
 
         if self._train_type in RECIPE_TRAIN_TYPE:
-            recipe = os.path.join(recipe_root, RECIPE_TRAIN_TYPE[self._train_type])
+            if self._train_type == TrainType.INCREMENTAL and self._hyperparams.learning_parameters.enable_supcon:
+                recipe = os.path.join(recipe_root, "supcon.py")
+                if "supcon" not in self._model_dir:
```
Comment (jaegukhyun) [inline otx/algorithms/segmentation/tasks/inference.py:192]: Is this check necessary? I think we always need to change our model directory to template_dir/supcon_model_dir

**Review by jaegukhyun — COMMENTED** [review]
Generally LGTM. Thank you for your hard work. I left some commands. 
It would be nice to re-check ci integration test from rebased with current develop branch

File: otx/algorithms/common/adapters/mmcv/hooks.py (hunk 2 above)
Comment (sungchul1) [reply otx/algorithms/common/adapters/mmcv/hooks.py:672]: > Just for curiosity, as far as I remember, the similar function was also implemented for Self-SL classification. Is it wrong? I just want to avoid the duplicate implementation

This hook is to handle interval (: when supcon loss is applied) in `TwoCropTransform` implemented in `otx/algorithms/segmentation/adapters/mmseg/data/pipelines.py` and this `TwoCropTransform` is similar with the other one implemented by adria in `otx/mpa/modules/datasets/pipelines/transforms/twocrop_transform.py`. There are obvious differences between both modules, but it is possible to combine them to the single module. But, this process is not a scope of this PR and it can be proceeded when merging mpa to otx.

Comment (sungchul1) [reply otx/algorithms/common/adapters/mmcv/hooks.py:672]: > Could you convert the doc-string format as Google's one? We'll use this format as standard doc-string format
> 
> Ref. https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

But other modules also use the same docstring format. I can modify it as you said, but it is not a scope of this PR and I think it would be better to modify it with other modules. What do you think?

File: otx/algorithms/common/tasks/training_base.py (hunk 4 above)
Comment (sungchul1) [reply otx/algorithms/common/tasks/training_base.py]: > @harimkang Is it okay to change this. AFAIK there is specific reason for this from building side

If so, I'll return it.

File: otx/algorithms/common/adapters/mmcv/hooks.py (hunk 2 above)
Comment (sungmanc) [reply otx/algorithms/common/adapters/mmcv/hooks.py:672]: > > Could you convert the doc-string format as Google's one? We'll use this format as standard doc-string format
> > Ref. https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html
> 
> But other modules also use the same docstring format. I can modify it as you said, but it is not a scope of this PR and I think it would be better to modify it with other modules. What do you think?

We're currently focusing on adding type-hint and doc-string. Although pre-existed code didn't include proper type-hint and doc-string, we suggested adding them to new PRs. So, please do not consider pre-existed code and modify just yours.

File: otx/algorithms/segmentation/tasks/inference.py (hunk 5 above)
Comment (sungchul1) [reply otx/algorithms/segmentation/tasks/inference.py:192]: I often occur the error that `self._model_dir` for training was remained until evaluation and `supcon` was added one more and `self._model_dir` became `.../supcon/supcon` whenever I ran the test.
SupCon for cls uses the similar condition to avoid overlapped path update.
https://github.com/openvinotoolkit/training_extensions/blob/8894d3d811ce1c770af3c1e3e74ec825a9ba16af/otx/algorithms/classification/tasks/inference.py#L380-L388
If there is another way to handle it, I can update it. Could you recommend it?

File: otx/algorithms/common/adapters/mmcv/hooks.py (hunk 3 above)
Comment (sungchul1) [reply otx/algorithms/common/adapters/mmcv/hooks.py]: Done.

File: otx/algorithms/classification/tasks/inference.py (hunk 1 above)
Comment (sungchul1) [reply otx/algorithms/classification/tasks/inference.py]: Done.

Comment (sungchul1) [reply otx/algorithms/classification/tasks/inference.py]:
```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

otx/algorithms/classification/tasks/inference.py:331: error: "None" has no attribute "get"  [attr-defined]
Found 1 error in 1 file (checked 626 source files)
```
It seems that `self._recipe_cfg` is `None` and mypy cannot call `get` method. Is there anything to resolve it?

Comment (sungmanc) [reply otx/algorithms/classification/tasks/inference.py]:
``` python
if self._recipe_cfg is not None:
    if params.enable_early_stopping is not None and self._recipe_cfg.get("evaluation", None) is not None:
        sth....***
```

Or, how about the below?

``` python
if self._recipe_cfg is not None:
    if params.enable_early_stopping is not None and self._recipe_cfg.get("evaluation") is not None:
        sth....***
```

Comment (supersoob) [reply otx/algorithms/classification/tasks/inference.py]: I think we need to add `assert self._recipe_cfg is not None` at L233 as defined in _init_recipe_hparam, or check 'if self._recipe_cfg is not None' here.

File: otx/algorithms/segmentation/tasks/inference.py (hunk 5 above)
Comment (supersoob) [reply otx/algorithms/segmentation/tasks/inference.py:192]: I glanced supcon support including classification, I feel like supcon is supported as train type already even if it has enable_supcon arg itself. (which has its own recipe file but the contents are almost same as incremental) I understand that model and data pipeline is totally different so it must be reloaded here but about other things, I personally recommend to refine all config for supcon in stage such as making _configure_supcon methods, injecting enable_supcon cfg in the incremental recipe file without making a recipe file. I think we need some discussion about that. But as of now, it's okay to handle these things (and consider classification supcon as well) after merging this PR.

File: otx/algorithms/classification/tasks/inference.py (hunk 1 above)
Comment (sungchul1) [reply otx/algorithms/classification/tasks/inference.py]: Ah, `self._recipe_cfg` is None during mypy only, not main execution.
So, we don't need to check if `self_recipe_cfg` is None.
I don't know the reason....

Comment (supersoob) [reply otx/algorithms/classification/tasks/inference.py]: I know, mypy checks the all the possibility that recipe cfg is none because it was defined as none initially. I think it's okay to add ignore: attr-defined for skipping mypy check.

Comment (sungmanc) [reply otx/algorithms/classification/tasks/inference.py]: > Ah, `self._recipe_cfg` is None during mypy only, not main execution. So, we don't need to check if `self_recipe_cfg` is None. I don't know the reason....

mypy consider all possible cases and doesn't like dynamic type changing. So, we need to explicit check like `if self._recipe_cfg is not None`. Then, mypy also knows `self._recipe_cfg` could not be None. 

Actually, in the training base (`BaseTask`), `self._recipe_cfg` is defined as None.
https://github.com/openvinotoolkit/training_extensions/blob/97e13ccf0275cbdc4005610e2c6c4259d0723d39/otx/algorithms/common/tasks/training_base.py#L106

File: otx/algorithms/common/adapters/mmcv/hooks.py (hunk 2 above)
Comment (sungchul1) [reply otx/algorithms/common/adapters/mmcv/hooks.py:672]: Done.

File: otx/algorithms/classification/tasks/inference.py (hunk 1 above)
Comment (sungchul1) [reply otx/algorithms/classification/tasks/inference.py]: Thanks for your kind explanation! @supersoob @sungmanc 
If adding `if self._recipe_cfg is not None` is more stable, I'll add it.

**Review by supersoob — DISMISSED** [review]
LGTM. Thanks for your hard work. I think it's okay to merge if all ci tests are passed. Let's discuss refactoring later.

**Review by sungmanc — DISMISSED** [review]

**Comment by sungchul1** [comment]
@supersoob @sungmanc There was some pylint issue and I fixed it. Could you approve it again?

**Review by supersoob — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]
